core/src/sumeragi/config.rs

```diff
@@ -0,0 +1,172 @@
+/// such as `BLOCK_TIME_MS` and list of `TRUSTED_PEERS`.
+#[derive(Debug, Clone, Default, PartialEq, Eq, Deserialize, Serialize)]
+#[serde(rename_all = "UPPERCASE")]
+#[serde(transparent)]
+pub struct TrustedPeers {
+    /// Optional list of predefined trusted peers. Must contain unique
+    /// entries. Custom deserializer raises error if duplicates found.
+    #[serde(deserialize_with = "deserialize_unique_trusted_peers")]
+    pub peers: HashSet<PeerId>,
+}
+
+/// Custom deserializer that ensures that `trusted_peers` only
+/// contains unique `PeerId`'s.
+///
+/// # Errors
+/// - Peer Ids not unique,
+/// - Not a sequence (array)
+fn deserialize_unique_trusted_peers<'de, D>(deserializer: D) -> Result<HashSet<PeerId>, D::Error>
+where
+    D: serde::Deserializer<'de>,
+{
+    /// Helper, for constructing a unique visitor that errors whenever
+    /// a duplicate entry is found.
+    struct UniqueVisitor(core::marker::PhantomData<fn() -> HashSet<PeerId>>);
+
+    impl<'de> serde::de::Visitor<'de> for UniqueVisitor {
+        type Value = HashSet<PeerId>;
+
+        fn expecting(&self, formatter: &mut core::fmt::Formatter) -> core::fmt::Result {
+            formatter.write_str("a set of unique `Peer::Id`s.")
+        }
+
+        fn visit_seq<S>(self, mut seq: S) -> Result<HashSet<PeerId>, S::Error>
+        where
+            S: serde::de::SeqAccess<'de>,
+        {
+            let mut result = HashSet::new();
+            while let Some(value) = seq.next_element()? {
+                if result.contains(&value) {
+                    return Err(serde::de::Error::custom(format!(
+                        "The peer id: {}'s public key appears twice.",
+                        &value
+                    )));
+                }
+                result.insert(value);
+            }
+
+            Ok(result)
+        }
+    }
+
+    let visitor = UniqueVisitor(core::marker::PhantomData);
+    deserializer.deserialize_seq(visitor)
+}
+
+impl TrustedPeers {
+    /// Load trusted peers variables from JSON.
+    ///
+    /// # Errors
+    /// - File not found
+    /// - File is not Valid JSON.
+    /// - File is valid JSON, but configuration options don't match.
+    pub fn from_path<P: AsRef<Path> + Debug>(path: P) -> Result<Self> {
+        let file = File::open(&path)
+            .wrap_err_with(|| format!("Failed to open trusted peers file {:?}", &path))?;
+        let reader = BufReader::new(file);
+        let trusted_peers: HashSet<PeerId> =
+            serde_json::from_reader(reader).wrap_err("Failed to deserialize json from reader")?;
+        Ok(TrustedPeers {
+            peers: trusted_peers,
+        })
+    }
+}
+
+fn default_peer_id() -> PeerId {
+    PeerId {
+        address: "".to_owned(),
+        public_key: PublicKey::default(),
+    }
+}
+
+// `HashSet::new()` is not const yet.
+fn default_empty_trusted_peers() -> TrustedPeers {
+    TrustedPeers {
+        peers: HashSet::new(),
+    }
+}
```

Comment:
Since the default for a primitive type is set, (and often zero), if we want the default to be anything but 0, we need a new type, with an `impl Default` for each. I'm not sure it's more compact that way.

Reply:
Every field is default so deriving should be enough though... am I missing something?

Reply:
Every field is default, but default is determined by type. So every field defaults to 0, which is not what we want.

Reply:
I don't get it.
None of `String`, `PublicKey`, `HashSet` has anything to do with the digit 0.
Also any other manual `impl Default` for these structs does not appear to me

Reply:
I think we may have misunderstood each other. Do you want me to remove `impl Default for SumeragiConfiguration` where these functions are used, or just to inline `default_peer_id`?

Reply:
Okay, my suggestion is as follows:

`iroha_data_model::peer`
```rust
#[derive(Default)]
pub struct Id {
    //
}
```
`iroha_core::sumeragi::config`
```rust
impl Default for SumeragiConfiguration {
            //
            trusted_peers: TrustedPeers::default(),
            peer_id: PeerId::default(),
            //
}

Reply:
Oh! Ok. That makes a lot of sense.